DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsDevice.h

```diff
@@ -27,15 +27,14 @@ class TrackingRecHitDevice : public PortableDeviceCollection<TrackingRecHitLayou
       : PortableDeviceCollection<TrackingRecHitLayout<TrackerTraits>, TDev>(nHits, queue), offsetBPIX2_{offsetBPIX2} {
     const auto device = alpaka::getDev(queue);
 
-    auto start_h = cms::alpakatools::make_host_view(hitsModuleStart, TrackerTraits::numberOfModules + 1);
+    auto start_h = cms::alpakatools::make_device_view(device, hitsModuleStart, TrackerTraits::numberOfModules + 1);
     auto start_d =
         cms::alpakatools::make_device_view(device, view().hitsModuleStart().data(), TrackerTraits::numberOfModules + 1);
     alpaka::memcpy(queue, start_d, start_h);
 
-    auto off_h = cms::alpakatools::make_host_view(offsetBPIX2);
+    auto off_h = cms::alpakatools::make_host_view(offsetBPIX2_);
```

Comment:
I'm afraid even this is not safe (unless the underlying GPU runtime+driver synchronize on the `memcpy()`). The `TrackingRecHitDevice` object (constructed in
https://github.com/cms-sw/cmssw/blob/28f76ff4318d495dce645342a4bc1554a0a59b66/RecoLocalTracker/SiPixelRecHits/plugins/alpaka/PixelRecHitKernels.dev.cc#L67
)
is moved around (the return from `makeHitsAsync()` may get elided, but in the end framework moves the object into the `edm::Wrapper<T>`), so the address of `offsetBPIX2_` is not stable.

But is `offsetBPIX2` really needed on the device? The only place I found was
https://github.com/cms-sw/cmssw/blob/28f76ff4318d495dce645342a4bc1554a0a59b66/RecoTracker/PixelSeeding/plugins/alpaka/CAHitNtupletGeneratorKernels.dev.cc#L285-L298
and everywhere else the `offsetBPIX2` is used on the host. In this case it would be easy to pass it by value from host (as the calling function has it already and uses it for other purposes). Or are there future developments that would benefit from having `offsetBPIX2` on the device as part of `TrackingRecHitsDevice`?

If `offsetBPIX2` is really needed on the device, I think it should be copied through a host buffer.

Reply:
Mhm.

The CUDA version does not have a synchronisation, either: https://github.com/cms-sw/cmssw/blob/a7047fdfa166c7546d46d9b21df0e1f5a2762d55/CUDADataFormats/TrackingRecHit/interface/TrackingRecHitSoADevice.h#L27-L46

Which is not to mean that I disagree with your assessment, just that the async copy seems to work anyway ?

(side note: for copying a small number of values, I'm wondering if using a kernel is not faster than calling a `cudaMemcpy` variant ?)

As to whether it's needed or not, I agree it needs to be looked into.

For the time being, I'm more interested in recovering the performance of the CUDA implementation, but after that we can also take the opportunity to clean up the code.

Reply:
I agree this PR is not worse than what is done in CUDA, so I'm not against merging it now (with the caveat that it is at least theoretically unsafe and should be eventually addressed).

> Which is not to mean that I disagree with your assessment, just that the async copy seems to work anyway ?

It could be the copy has been so quick that it has completed before the `memcpy()` has finished (really along "nothing else has overwritten the memory before CUDA runtime+driver read from the memory location"). I'd expect the problem manifest when "the GPU is at least nearly full", and even then it is unclear whether it would crash the job or lead to incorrect value to be read (and whatever that would imply downstream).

Reply:
> It could be the copy has been so quick that it has completed before the `memcpy()` has finished (really along "nothing else has overwritten the memory before CUDA runtime+driver read from the memory location").

Actually, the CUDA documentation is not very clear what "async memcpy" means:
  - for a _synchronous_ copy it states that
    > For transfers from pageable host memory to device memory, a stream sync is performed before the copy is initiated. The function will return once the pageable buffer has been copied to the staging memory for DMA transfer to device memory, but the DMA to final destination may not have completed. 
  - for an _asynchronous_ copy it states that
    > If pageable memory must first be staged to pinned memory, the driver may synchronize with the stream and stage the copy into pinned memory. 
    
So it's possible that the CUDA runtime may first perform a synchronous copy of `offsetBPIX2_` to a staging area in pinned memory, and then an asynchronous copy from there to GPU memory.

Or not 🤷🏻‍♂️ .

Reply:
Right, the exact behavior is confusing, and I'd expect to depend on the runtime and driver versions and/or actual hardware. I'd personally assume the weakest guarantees and program around that.

(and with Alpaka we should worry about the behavior of other platforms too)